hoomd/hpmc/IntegratorHPMCMonoGPU.h

```diff
@@ -769,303 +1115,467 @@ void IntegratorHPMCMonoGPU< Shape >::update(unsigned int timestep)
+                                    work_offset[idev] = 0;
+                                    }
+
+                                #ifdef ENABLE_MPI
+                                if (m_ntrial_comm)
+                                    {
+                                    // split up work among ranks
+                                    for (unsigned int idev = 0; idev < this->m_exec_conf->getNumActiveGPUs(); ++idev)
+                                        {
+                                        nwork_rank[idev] = nwork_rank[idev]/ntrial_comm_size + 1; // can't have zero work per rank
+                                        work_offset[idev] = ntrial_comm_rank*nwork_rank[idev];
+                                        }
+                                    }
+                                #endif
+
+                                gpu::hpmc_auxilliary_args_t auxilliary_args(
+                                    d_tag.data,
+                                    d_vel.data,
+                                    d_trial_vel.data,
+                                    ntrial,
+                                    &nwork_rank[0],
+                                    &work_offset[0],
+                                    d_n_depletants_ntrial.data + ntrial_offset,
+                                    d_deltaF_int.data + this->m_depletant_idx(itype,jtype)*this->m_pdata->getMaxN(),
+                                    &m_depletant_streams_phase1[this->m_depletant_idx(itype,jtype)].front(),
+                                    &m_depletant_streams_phase2[this->m_depletant_idx(itype,jtype)].front(),
+                                    m_max_len,
+                                    d_req_len.data,
+                                    particle_comm_rank == particle_comm_size - 1,
+                                    this->m_pdata->getNGhosts(),
+                                    gpu_partition_rank);
+
+                                // phase 1, insert into excluded volume of particle i
+                                m_tuner_depletants_phase1->begin();
+                                unsigned int param = m_tuner_depletants_phase1->getParam();
+                                args.block_size = param/1000000;
+                                implicit_args.depletants_per_thread = (param % 1000000)/10000;
+                                args.tpp = param%10000;
+                                gpu::hpmc_depletants_auxilliary_phase1<Shape>(args,
+                                    implicit_args,
+                                    auxilliary_args,
+                                    params.data());
+                                if (this->m_exec_conf->isCUDAErrorCheckingEnabled())
+                                    CHECK_CUDA_ERROR();
+                                m_tuner_depletants_phase1->end();
+
+                                // phase 2, reinsert into excluded volume of i's neighbors
```

Comment:
It looks like you're trying to understand the method. And to thoroughly explain it, I admit I really have to write a paper, otherwise from just looking at the code it may seem to come out of nowhere :-)

Let me try to answer your question. The idea of the algorithm is that a Poisson point process provides an unbiased estimator of the Boltzmann weight for the overlap volume, and every realization of such a process represents a microstate, the ensemble of which we sample through the usual MCMC (Markov Chain Monte Carlo).

To understand the need for the two kernels, note that if we're moving particle i, we are also proposing a new configuration of test particles overlapping with it  (stored compactly as a single random number generator seed in the auxiliary variable). The configuration of this particle and its test particles then represent the new microstate, and we compute the Metropolis-Hastings (MH) weight as the usual ratio of estimators (**phase 1**). Because other particles `j!=i` and their associated depletants remain in their existing configurations, we have to account for the possibility that i's move changes the outcome of its neighbor j's estimators. In particular, if i overlaps with any of j's test particles before or after the move, we have to include the new and the old estimate for j in both the numerator and the denominator of the acceptance probability, by re-inserting test particles for j-neighbors whenever we move i (**phase 2**).

Looking at the serial implementation you'll notice that we perform this reinsertion of test particles for neighbors with every trial move  - the number of depletant insertions therefore scales as `O(N_neigh)*O(N)`, which is slower than the `O(N)` of the simple rejection algorithm (in addition to the factor of two for evaluating numerator and denominator separately). Of course, there are still `N_avg` test particles being inserted per actual particle. The split into two kernels reduces the number of reinsertions back to `O(N)*N_avg` (on the GPU), by inserting all i test particles in phase 1, and by simultaneously inserting into all j-neighbors in phase 2 (the phases do not depend on each other). More specifically, in phase 2 kernel a j particle touches i's memory, where as in phase 1 the mapping between particles and acceptance probabilities is the identity. The slight implication of this access pattern (for multi-GPU performance) is that the random access from j -> i in phase 2 necessitates page migrations, if particles i and j are on different GPUs.

Reply:
Yes, a paper would be nice :) but your logic makes sense, I see what you're doing. Thanks for the explanation!